### DISCUSSION TITLE
release schedule

### DISCUSSION THREAD

**Discussion by pwoolvett** [flakeheaven/flakeheaven#50] [Q&A]
In the mid term, Id like to setup conventional commits + commitizen for auto release on merges to main branch.

The core idea is to:

0. disallow main push (if not already disabled)
1. only allow squash pr merges
2. require conventional commits 
3. once pr is accepted, tools (eg commitizen) can automatically generate changelog and bump major/minor/patch automatically
4. add the bump into ci and release automatically to pypi.

In the mean time, I've been releasing manually as I see fit.

Any volunteers or alternatives before implementing proposed methods?

**Comment by pwoolvett** [comment]
@jarojasm95 finished writing this and saw [your comment](https://github.com/flakeheaven/flakeheaven/pull/47#pullrequestreview-874092675)

**Comment by jose-rojas0** [comment]
releasing on _every_ push to `main` branch might be too much? although that config would likely be the easiest to setup

the proposal sounds like a good starting point, we can always iterate on it as we see the need, do you have an example of what this workflow would look like? (in a different repo that uses it perhaps?)

**Reply by pwoolvett** [reply]
> releasing on _every_ push to `main` branch might be too much? although that config would likely be the easiest to setup

Not on _every_ push to `main`. Only three kinds of commits produce a new release using this schema: 

* major: breaking changes, dropping support, etc. Only commits with either a `BREAKING CHANGE` trailing commit msg or with a `!` in its header produce this.
* minor: new features. Only commits with a `feat` in its header produce this.
* patch: end-user related fixes (not ci/build/etc). Only commits with a `fix` in its header produce this.

final version is MAJOR.MINOR.PATCH.

Tools can enforce commits comply with the spec in PRs.

> the proposal sounds like a good starting point, we can always iterate on it as we see the need, do you have an example of what this workflow would look like? (in a different repo that uses it perhaps?)

On private repos Ive set it up like this from reference implementation [here](https://github.com/commitizen-tools/commitizen-action)

So basically, :

* `ci.yaml`
  - trigger: PRs
  - "normal" ci: what we have now, but only for PRs.
  - add a conventional commit check to enforce compliance
* `bumpversion.yaml`: "Bump version on main push" -- 
  - trigger: `main` push (accepted pr are `main` push as well...)
  - Only if at least one of the three mentioned conditions are met, conditionally tags, bumps, and pushes tags to repo.
  - linked github action/workflow reference
* `publish.yaml`:
  - trigger: tags only
  - should run `poetry publish --build` etc

**Reply by jose-rojas0** [reply]
this sounds great to me.

**Reply by mcarans** [reply]
Sounds good to me too.

**Reply by pwoolvett** [reply]
https://github.com/flakeheaven/flakeheaven/pull/92 please review / accept and we can move on!

slight change with respect to this spec: we;d be using rebase instead of squash and keep every commit. CON: a liitle bit more involved for contributors not used to squash/rebase their own PRs. PRO: less work for maintainers, cleaner timeline.
